src/lib.rs

```diff
@@ -39,12 +42,222 @@ pub enum Error {
     /// The range that we try to insert into the interval tree is overlapping
     /// with another node from the tree.
     #[error("Addresses are overlapping.{0:?} intersects with existing {1:?}")]
-    Overlap(allocation_engine::Range, allocation_engine::Range),
+    Overlap(Range, Range),
     /// A node state can be changed just from Free to Allocated, other transitions
     /// are not valid.
     #[error("Invalid state transition for node {0:?} from {1:?} to NodeState::Free")]
-    InvalidStateTransition(allocation_engine::Range, allocation_engine::NodeState),
+    InvalidStateTransition(Range, allocation_engine::NodeState),
+    /// Address is unaligned
+    #[error("The address is not aligned.")]
+    UnalignedAddress,
 }
 
 /// Wrapper over std::result::Result
 pub type Result<T> = result::Result<T, Error>;
+
+/// A closed interval range [start, end] used to describe a memory slot that
+/// will be assigned to a device by the VMM. This structure represents the key
+/// of the Node object in the interval tree implementation.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord, Debug)]
+pub struct Range {
+    /// Lower boundary of the interval.
+    start: u64,
+    /// Upper boundary of the interval.
+    end: u64,
+}
+
+#[allow(clippy::len_without_is_empty)]
+impl Range {
+    /// Create a new Range object.
+    pub fn new(start: u64, end: u64) -> Result<Self> {
+        if start > end {
+            return Err(Error::InvalidRange(start, end));
+        }
+        Ok(Range { start, end })
+    }
+
+    /// Get length of the range.
+    pub fn len(&self) -> u64 {
+        self.end - self.start + 1
+    }
+
+    /// Check whether two Range objects overlap with each other.
+    pub fn overlaps(&self, other: &Range) -> bool {
+        max(self.start, other.start) <= min(self.end, other.end)
+    }
+
+    /// Check whether the key is fully covered.
+    pub fn contains(&self, other: &Range) -> bool {
+        self.start <= other.start && self.end >= other.end
+    }
+
+    /// Get the lower boundary of the range.
+    pub fn get_start(&self) -> u64 {
```

Comment:
style: With a few exceptions, the `get_` prefix is not used for getters in Rust code. [source](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter)
Same for `get_end()`.